File: handlers/pjsk/asset_binary.go
package pjsk

import (
	"encoding/base64"
	"fmt"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
)

const assetBinaryCacheControl = "public, max-age=31536000, immutable"

func decodeDataURL(dataURL string) (string, []byte, error) {
	raw := strings.TrimSpace(dataURL)
	if !strings.HasPrefix(raw, "data:") {
		return "", nil, fmt.Errorf("invalid data URL prefix")
	}

	metaAndPayload := strings.TrimPrefix(raw, "data:")
	meta, payload, ok := strings.Cut(metaAndPayload, ",")
	if !ok || payload == "" {
		return "", nil, fmt.Errorf("invalid data URL payload")
	}
	if !strings.HasSuffix(meta, ";base64") {
		return "", nil, fmt.Errorf("unsupported data URL encoding")
	}

	contentType := strings.TrimSuffix(meta, ";base64")
	if contentType == "" {
		contentType = "application/octet-stream"
	}

	data, err := base64.StdEncoding.DecodeString(payload)
	if err != nil {
		return "", nil, fmt.Errorf("decode base64 failed: %w", err)
	}
	return contentType, data, nil
}

// AssetBinaryHandler GET /pjsk/assets/:label
// 示例: /pjsk/assets/event:background:event_marathon_001?server=jp
func AssetBinaryHandler(c *gin.Context) {
	server := c.DefaultQuery("server", "jp")
	if !validServers[server] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "无效的服务器参数，支持: jp, cn, en, tw, kr"})
		return
	}

	label := strings.TrimSpace(c.Param("label"))
	if label == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "缺少资源 label"})
		return
	}

	if normalizedLabel, _ := buildRelativePathsByLabel(label); normalizedLabel == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "不支持的资源 label: " + label})
		return
	}

	dataURL := downloadAssetByLabel(server, label)
	if dataURL == "" {
		c.JSON(http.StatusNotFound, gin.H{"error": "资源下载失败: " + label})
		return
	}

	contentType, payload, err := decodeDataURL(string(dataURL))
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "解析资源失败: " + err.Error()})
		return
	}

	c.Header("Cache-Control", assetBinaryCacheControl)
	c.Data(http.StatusOK, contentType, payload)
}

File: src/handlers/pjsk/asset_binary.rs
use crate::handlers::pjsk::asset_source::{build_relative_paths_by_label, download_asset_by_label};
use crate::handlers::pjsk::VALID_SERVERS;
use axum::{
    extract::{Path as AxumPath, Query},
    http::{header, StatusCode},
    response::IntoResponse,
};
use serde::Deserialize;

const ASSET_BINARY_CACHE_CONTROL: &str = "public, max-age=31536000, immutable";

#[derive(Deserialize, Debug)]
pub struct AssetBinaryQuery {
    pub server: Option<String>,
}

pub fn decode_data_url(data_url: &str) -> Result<(String, Vec<u8>), String> {
    let raw = data_url.trim();
    if !raw.starts_with("data:") {
        return Err("invalid data URL prefix".to_string());
    }

    let meta_and_payload = &raw["data:".len()..];
    let (meta, payload) = meta_and_payload
        .split_once(',')
        .ok_or_else(|| "invalid data URL payload".to_string())?;

    if !meta.ends_with(";base64") {
        return Err("unsupported data URL encoding".to_string());
    }

    let content_type = &meta[..meta.len() - ";base64".len()];
    let content_type = if content_type.is_empty() {
        "application/octet-stream"
    } else {
        content_type
    };

    use base64::Engine;
    let data = base64::engine::general_purpose::STANDARD
        .decode(payload)
        .map_err(|e| format!("decode base64 failed: {}", e))?;

    Ok((content_type.to_string(), data))
}

pub async fn asset_binary_handler(
    AxumPath(label): AxumPath<String>,
    Query(q): Query<AssetBinaryQuery>,
) -> impl IntoResponse {
    let server = q.server.unwrap_or_else(|| "jp".to_string());
    if !VALID_SERVERS.contains(&server) {
        return (
            StatusCode::BAD_REQUEST,
            "无效的服务器参数，支持: jp, cn, en, tw, kr",
        )
            .into_response();
    }

    let label = label.trim();
    if label.is_empty() {
        return (StatusCode::BAD_REQUEST, "缺少资源 label").into_response();
    }

    let (normalized, _) = build_relative_paths_by_label(label);
    if normalized.is_empty() {
        return (
            StatusCode::BAD_REQUEST,
            format!("不支持的资源 label: {}", label),
        )
            .into_response();
    }

    let data_url = download_asset_by_label(&server, label).await;
    if data_url.is_empty() {
        return (StatusCode::NOT_FOUND, format!("资源下载失败: {}", label)).into_response();
    }

    match decode_data_url(&data_url) {
        Ok((content_type, payload)) => (
            StatusCode::OK,
            [
                (header::CACHE_CONTROL, ASSET_BINARY_CACHE_CONTROL),
                (header::CONTENT_TYPE, content_type.as_str()),
            ],
            payload,
        )
            .into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            format!("解析资源失败: {}", e),
        )
            .into_response(),
    }
}
